src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs

```diff
@@ -149,32 +148,34 @@ private static unsafe void CreateSocket()
                 throw new NetworkInformationException(message);
             }
 
-            s_socket = newSocket;
-            new Thread(s => LoopReadSocket((int)s!))
+            s_socket = new SocketWrapper(newSocket);
+
+            new Thread(args => LoopReadSocket((SocketWrapper)args!))
             {
                 IsBackground = true,
                 Name = ".NET Network Address Change"
-            }.UnsafeStart(newSocket);
+            }.UnsafeStart(s_socket);
         }
 
         private static void CloseSocket()
         {
-            Debug.Assert(s_socket != 0, "s_socket was 0 when CloseSocket was called.");
-            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket);
+            Debug.Assert(s_socket != SocketWrapper.NotSet, "s_socket was not set when CloseSocket was called.");
+            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket.Socket);
             if (result != Interop.Error.SUCCESS)
             {
                 string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
                 throw new NetworkInformationException(message);
             }
 
-            s_socket = 0;
+            s_socket = SocketWrapper.NotSet;
         }
 
-        private static unsafe void LoopReadSocket(int socket)
+        private static unsafe void LoopReadSocket(SocketWrapper initiallyCreatedSocket)
         {
-            while (socket == s_socket)
+            while (s_socket == initiallyCreatedSocket) //if both references are equal then s_socket was not changed
             {
-                Interop.Sys.ReadEvents(socket, &ProcessEvent);
+                //we can continue processing events
+                Interop.Sys.ReadEvents(initiallyCreatedSocket.Socket, &ProcessEvent);
```

Comment:
Between the condition-check in the while loop, and passing the handle to `ReadEvents` the handle may be re-used for something other. That means we'll be reading for some one else's file descriptor.

Using `Socket`/`SafeSocketHandle` allows to fix that.

Reply:
That may be bigger change than  @vdjuric may want to take.  If we feel that would simplify the code, I think we should open separate issue for it @tmds. 
Alternatively, we can possibly use `SafeHandle` but that would also be much bigger change IMHO.